src/components/Time.tsx

```diff
@@ -1,49 +1,49 @@
 import './styles/time.less'
 
 import React from 'react'
-import PropTypes from 'prop-types'
 
 import Tooltip from './Tooltip'
-
-import TimeUtils from './utils/TimeUtils'
+import TimeUtils, { TimeConfig } from './utils/TimeUtils'
 
 const DELAY = 60000
 
-/**
- * @typedef {Object} Props
- * @property {number} time
- * @property {import('./utils/TimeUtils').Config} config
- * @property {string} [className]
- * @property {boolean} [tooltipOnHover]
- * @property {boolean} [refresh]
- * @property {React.ReactNode} [tooltip]
- * @property {boolean} [expandOnHover]
- * @property {boolean} [expandFullyOnHover]
- */
-
-/**
- * @typedef {Object} State
- * @property {boolean} hovered
- */
-
-/** @extends {React.Component<Props, State>} */
-class Time extends React.Component<any, any> {
-    public static displayName
-    public static propTypes
-    public static defaultProps
-
-    /**
-     * @param {Props} props
-     */
-    constructor(props) {
+interface Props {
+    /** UNIX timestamp of the time to display. */
+    time?: number
+    /** Configuration for localization. */
+    config?: TimeConfig
+    /** Additional css class applied to the time element. */
+    className?: string
+    tooltipOnHover?: boolean
+    /** Refresh the component every DELAY seconds. */
+    refresh?: boolean
+    /** If you don't want to use the default time format on the tooltip use this prop to supply a custom text */
+    tooltip?: React.ReactNode
+    /** When hovering over time it expands to short absolute version. */
+    expandOnHover?: boolean
+    /** When hovering over time it expands to the full absolute version. */
+    expandFullyOnHover?: boolean
+}
+
+interface State {
+    hovered: boolean
+    mouseX?: number
+    mouseY?: number
+}
+
+class Time extends React.Component<Props, State> {
+    public static displayName: string
+    public static defaultProps: Props
+
+    constructor(props: Props) {
         super(props)
         //eslint-disable-next-line @typescript-eslint/camelcase
         ;(this as any).refresh_interval = null
 
         this.state = {
             hovered: false,
-            mouseX: null,
-            mouseY: null,
+            mouseX: undefined,
```

Comment:
@gnapse @henningmu Worth discussing as element of code style. Personally I think that `undefined` on object properties is confusing and it is better to use `null` in such cases. I would recomment having `mouseX: number | null` instead.

Reply:
Yes I agree that we should pick one and stick to it. I prefer `undefined` because it avoids an extra `| null` suffix in the type.

```
interface State {
    hovered: boolean
    mouseX: number | null
    mouseY: number | null
}
```

when there is great syntactic sugar available with the `?: number`. JS is kinda terrible in that it has 2 types of nulls 😶

Reply:
The problem with undefined is that there's no distinction between type of `foo` in `{}` and `{foo: undefined}`. In a case like `State` above it loses all advantage of static typing (you could forget the property). My recommendation would be to never use undefined as explicit type (only use it to mark optionality of param).

Reply:
I consider this a good thing 🙂 If you use undefined throughout then you've eliminated 2 classes of bugs here: the property not being present, and the property being undefined, and your type guards take care of the third one, the property being null. 

This is really stylistic preference though, but for what it's worth (& the big warnings), Typescript's code itself is [written to prefer undefined](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined).

Reply:
I don‘t have a strong opinion to this, you both made great arguments. At first, I was somehow convinced by the use of `null`, like all the other languages does. But @vishnugopal provides good arguments for the use of `undefined` and it shifted my mind, so I would vote for `undefined`.

Reply:
@vishnugopal Your arguments are solid, let's stay with undefined.

Reply:
Didn't get the chance to chime in, but Vishnu covered really well what I would've argued for as well. Even better, as I wouldn't have been able to express it so well.